barretenberg/cpp/src/barretenberg/world_state/world_state.cpp

```diff
@@ -353,24 +361,26 @@ bool WorldState::sync_block(StateReference& block_state_ref,
 {
     Fork::SharedPtr fork = retrieve_fork(CANONICAL_FORK_ID);
     auto current_state = get_state_reference(WorldStateRevision::uncommitted());
-    if (block_state_matches_world_state(block_state_ref, current_state)) {
-        // TODO (alexg) remove commit & rollback. All modifications should happen on forks
-        // Synching a block should always be done on top of a clean state
-        // so committing partial writes like we're doing here shouldn't be possible
-        //
-        // Disable updating the archive tree. Why? The only way to commit dirty state when synching a block is if the
-        // current node _built_ the block. If that's the case then the archive tree has already been updated by the
-        // node's orchestrator.
-        // append_leaves<fr>(MerkleTreeId::ARCHIVE, { block_hash });
-        // just commit the dirty state
+    if (block_state_matches_world_state(block_state_ref, current_state) &&
+        is_archive_tip(WorldStateRevision::uncommitted(), block_header_hash)) {
         commit();
         return true;
     }
 
     rollback();
 
     Signal signal(static_cast<uint32_t>(fork->_trees.size()));
-    auto decr = [&signal](const auto&) { signal.signal_decrement(); };
+    bool success = true;
```

Comment:
This will need to be std::atomic_bool and you will need to do a compare exchange when testing and setting. e.g.

```
bool temp = true;
if (!resp.success && success.compare_exchange_strong(temp, false))
{
}
```